### ISSUE TITLE
missing_safety_doc should be surpressed by parent #[doc(hidden)]

### ISSUE THREAD

**Issue by CAD97** [rust-lang/rust-clippy#7347]
Lint name: missing_safety_doc 

I tried [this code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b727898231f427fa5ffe23fa4e373634):

```rust
//!
#![warn(missing_docs)]

#[doc(hidden)]
pub mod __macro {
    pub struct T;
    impl T {
        pub unsafe fn f() {}
    }
}
```

I expected to see this happen: no lints fire

Instead, this happened: clippy::missing_safety_doc fired

`#[doc(hidden)]` on an item suppresses `clippy::missing_safety_doc`. However, while `#[doc(hidden)]` on a parent module suppresses `missing_docs`, it does not suppress `clippy::missing_safety_doc`. It's probably better to match the behavior of `missing_docs` here, even though `clippy::missing_safety_doc` fires for private items, and `missing_docs` doesn't.

### Meta

- `cargo clippy -V`: clippy 0.1.53 (4bac69dd250 2021-05-07)
- `rustc -Vv`:
  ```
  rustc 1.53.0-beta.2 (4bac69dd2 2021-05-07)
  binary: rustc
  commit-hash: 4bac69dd25089dd2f803f24de2225d84449b6220
  commit-date: 2021-05-07
  host: x86_64-pc-windows-msvc
  release: 1.53.0-beta.2
  LLVM version: 12.0.0
  ```
